doc/sphinx/source/users-guide/base/zip.rst

```diff
@@ -0,0 +1,133 @@
+:title: Users Guide: Zippered Iteration
+
+.. _ug-zip:
+
+Zippered Iteration
+==================
+
+Chapel's :ref:`for-loops <ug-forloops>` (and their parallel
+counterparts, :ref:`coforall <ug-coforall>` and forall) can be driven
+by multiple iterands in a coordinated manner known as *zippered
+iteration*.  This is expressed using the ``zip`` keyword, followed by
+a parenthesized list of the iterands.
+
+As a simple example, the following loop iterates over an array and a
+range simultaneously, referring to the values yielded by each using
+distinct index variables, *a* and *i*:
+
+.. literalinclude:: examples/users-guide/base/zip.chpl
+  :caption:
+  :language: chapel
+  :lines: 1-3
+
+As the name suggests, zippered iteration causes the loop body to be
+executed once per corresponding set of values yielded by the iterands.
+Thus, in the first iteration of the loop above, *a* will refer to the
+first element of *A* and *i* to 1; in the second, *a* refers to *A*'s
+second element and *i* to 2; etc.  For this reason, zippered iterands
+must typically have the same "size" (yield the same number of things),
+though we'll discuss exceptional cases below.
+
+As a result of this behavior, printing the array after this loop has
+executed will result in:
+
+.. literalinclude:: examples/users-guide/base/zip.good
+  :language: text
+  :lines: 1
+
+Naturally, zippering may involve more than two iterands as well:
+
+.. literalinclude:: examples/users-guide/base/zip.chpl
+  :language: chapel
+  :lines: 7-8
+
+Here, the array ends up as:
+
+.. literalinclude:: examples/users-guide/base/zip.good
+  :language: text
+  :lines: 3
+
+
+
+Index Variables for Zippered Iteration
+--------------------------------------
+
+In the example above, we captured the results of the zippering using
+an index variable per iterand.  Zippered iterations can also be
+represented using a single index variable storing a tuple of the
+iterands' values:
+
+.. literalinclude:: examples/users-guide/base/zip.chpl
+  :language: chapel
+  :lines: 12-13
+
+The output of this loop is:
+
+.. literalinclude:: examples/users-guide/base/zip.good
+  :language: text
+  :lines: 5-9
+
+In fact, the original case can be thought of as de-tupling the
```

Comment:
Could also say "de-structuring" or "pattern matching", but maybe better left for "Chapel for Functional Programmers".